### PR TITLE
Fix for HypreParMatrixFromBlocks for square matrices

### PR DESCRIPTION
**Pull request by sebastiangrimberg**
If using `HypreParMatrixFromBlocks` to form a square matrix, we should call the `HypreParMatrix` constructor with `row_starts = col_starts` so that `hypre_CSRMatrixReorder` gets properly called. Without this, methods like `HypreParMatrix::GetDiag` will fail on the merged matrix.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2804](https://github.com/mfem/mfem/pull/2804) | @sebastiangrimberg | @pazner | @dylan-copeland + @jandrej | 2/4/22 | 3/22/22 | ⌛due 2/25/22 | |

### CODE REVIEW

**Comment by pazner** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: linalg/hypre.cpp
Comment (dylan-copeland) [inline linalg/hypre.cpp:2980]: It seems very unlikely, but possible, that a rectangular matrix could have `all_num_loc_rows[rank] == all_num_loc_cols[rank]` for a particular rank but not for other ranks. Would it be better to loop over all ranks and check this condition, to set a flag `bool square`?

**Comment by dylan-copeland** [comment]
There was a failing test, `mpirun -np 2 ex0p`, on macos-10.15-make-opt-par-int32, with a seg fault. The test works on my mac. @pazner could this have been an issue with the testing process, so we should restart the tests?

**Comment by pazner** [comment]
> There was a failing test, `mpirun -np 2 ex0p`, on macos-10.15-make-opt-par-int32, with a seg fault. The test works on my mac. @pazner could this have been an issue with the testing process, so we should restart the tests?

I restarted the tests and they are passing now, hopefully it was some kind of CI glitch.

File: linalg/hypre.cpp
Comment (sebastiangrimberg) [reply linalg/hypre.cpp:2980]: That's right, I guess my point was just that even if the global matrix isn't square, doing the diagonal reordering locally doesn't seem like it would cause any problem and avoids a global communication to check that every rank is square.

Comment (dylan-copeland) [reply linalg/hypre.cpp:2980]: That may be a valid point about the local diagonal reordering, but another issue is that passing the same pointer for `rows` and `cols` to the 9-argument constructor will change the behavior of the constructor. These arrays are already of size equal to the number of ranks, and the communication is already done by `GatherBlockOffsetData`, so the cost to check all ranks is negligible. More important is to prevent going into the square case of the constructor on some ranks and the rectangular case on other ranks, which seems sure to cause problems.

Comment (sebastiangrimberg) [reply linalg/hypre.cpp:2980]: Yeah, that sounds fair enough. I’ll add the check in ASAP.

File: linalg/hypre.cpp
Comment (dylan-copeland) [inline linalg/hypre.cpp:2981]: This looks correct, but I think you can avoid the `MPI_Allreduce` since the sizes are already gathered by `GatherBlockOffsetData` into `all_num_loc_rows` and `all_num_loc_cols`, which are arrays with sizes for all ranks. Also, `first_loc_row` and `first_loc_col` are computed from these arrays, so there is no need to compare them. You can just loop over all ranks `r` and check `all_num_loc_rows[r] == all_num_loc_cols[r]`.

Comment (sebastiangrimberg) [reply linalg/hypre.cpp]: Yup, duh. Should be fixed now.

**Review by dylan-copeland — APPROVED** [review]
Thanks for the fix, @sebastiangrimberg!

**Comment by pazner** [comment]
@jandrej, could you take a look when you get a moment?

**Comment by pazner** [comment]
@jandrej, could you take a look when you have the time? It would be good to merge this once it's approved.

**Review by jandrej — APPROVED** [review]

**Comment by pazner** [comment]
Merged in `next` for testing...

**Pull request merged**
